rust/noosphere/src/ffi/fs.rs

```diff
@@ -0,0 +1,233 @@
+use anyhow::anyhow;
+use noosphere_core::data::Did;
+use noosphere_fs::{SphereFile, SphereFs};
+use safer_ffi::prelude::*;
+use std::{pin::Pin, str::FromStr};
+use subtext::{Peer, Slashlink};
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+use crate::{
+    ffi::{NsHeaders, NsNoosphereContext},
+    platform::{PlatformKeyMaterial, PlatformStore},
+};
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFs {
+        inner: SphereFs<PlatformStore, PlatformKeyMaterial>
+    }
+}
+
+impl NsSphereFs {
+    pub fn inner(&self) -> &SphereFs<PlatformStore, PlatformKeyMaterial> {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut SphereFs<PlatformStore, PlatformKeyMaterial> {
+        &mut self.inner
+    }
+}
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFile {
+        inner: SphereFile<Pin<Box<dyn AsyncRead>>>
+    }
+}
+
+impl NsSphereFile {
+    pub fn inner(&self) -> &SphereFile<Pin<Box<dyn AsyncRead>>> {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut SphereFile<Pin<Box<dyn AsyncRead>>> {
+        &mut self.inner
+    }
+}
+
+#[ffi_export]
+/// Initialize an instance of a [NsSphereFs] that is a read/write view into
+/// the contents (as addressed by the slug namespace) of the identitifed sphere
+pub fn ns_sphere_fs_open(
+    noosphere: &NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+) -> repr_c::Box<NsSphereFs> {
+    noosphere
+        .async_runtime()
+        .block_on(async {
+            let sphere_context = noosphere
+                .inner()
+                .get_sphere_context(&Did(sphere_identity.to_str().into()))
+                .await?;
+
+            let sphere_context = sphere_context.lock().await;
+
+            Ok(repr_c::Box::new(NsSphereFs {
+                inner: sphere_context.fs().await?,
+            })) as Result<_, anyhow::Error>
+        })
+        .unwrap()
+}
+
+#[ffi_export]
+/// De-allocate an [NsSphereFs] instance
+pub fn ns_sphere_fs_free(sphere_fs: repr_c::Box<NsSphereFs>) {
+    drop(sphere_fs)
+}
+
+#[ffi_export]
+/// Read a [NsSphereFile] from a [NsSphereFs] instance by slashlink. Note that
+/// although this function will eventually support slashlinks that include the
+/// pet name of a peer, at this time only slashlinks with slugs referencing the
+/// slug namespace of the local sphere are allowed.
+pub fn ns_sphere_fs_read(
```

Comment:
As you can see from the implementation, `read` is the literal name of the method of `SphereFs` being invoked. I'm inclined to keep it because when reading a file in almost any FS module, you will use some implementation that uses the verb `read`.

`open` on `SphereFs` was chosen somewhat arbitrarily. Maybe it would be better to change that?

Reply:
`open` on `NsSphereFs` makes sense to me -- keeping `read` aligned with the underlying `SphereFs` method also makes sense, with the laziness of `NsSphereFile` being an implementation detail